File: irrexplorer/backends/rdap.py
"""
RDAP (Registration Data Access Protocol) backend integration.

Provides domain and IP registration data from RDAP servers.
"""

import asyncio
import logging
from typing import Any, Dict, Optional
from urllib.parse import quote

import aiohttp

logger = logging.getLogger(__name__)


class RDAPClient:
    """Client for RDAP queries."""

    # Bootstrap servers for different RIRs
    BOOTSTRAP_SERVERS = {
        "arin": "https://rdap.arin.net/registry",
        "ripe": "https://rdap.db.ripe.net",
        "apnic": "https://rdap.apnic.net",
        "lacnic": "https://rdap.lacnic.net/rdap",
        "afrinic": "https://rdap.afrinic.net/rdap",
    }

    def __init__(self, timeout: int = 30):
        """
        Initialize RDAP client.

        Args:
            timeout: Request timeout in seconds
        """
        self.timeout = aiohttp.ClientTimeout(total=timeout)

    async def query_ip(
        self, ip_address: str, rir: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Query RDAP for IP address registration data.

        Args:
            ip_address: IP address to query
            rir: Optional specific RIR to query (arin, ripe, etc.)

        Returns:
            Dictionary with registration data
        """
        if rir and rir.lower() in self.BOOTSTRAP_SERVERS:
            return await self._query_rir(ip_address, rir.lower(), "ip")

        # Try all RIRs if no specific one provided
        for rir_name in self.BOOTSTRAP_SERVERS:
            result = await self._query_rir(ip_address, rir_name, "ip")
            if result and not result.get("error"):
                return result

        return {"ip": ip_address, "error": "Not found in any RIR"}

    async def query_asn(self, asn: int, rir: Optional[str] = None) -> Dict[str, Any]:
        """
        Query RDAP for ASN registration data.

        Args:
            asn: Autonomous System Number
            rir: Optional specific RIR to query

        Returns:
            Dictionary with registration data
        """
        if rir and rir.lower() in self.BOOTSTRAP_SERVERS:
            return await self._query_rir(str(asn), rir.lower(), "autnum")

        # Try all RIRs if no specific one provided
        for rir_name in self.BOOTSTRAP_SERVERS:
            result = await self._query_rir(str(asn), rir_name, "autnum")
            if result and not result.get("error"):
                return result

        return {"asn": asn, "error": "Not found in any RIR"}

    async def query_domain(self, domain: str) -> Dict[str, Any]:
        """
        Query RDAP for domain registration data.

        Args:
            domain: Domain name to query

        Returns:
            Dictionary with registration data
        """
        try:
            # Use ICANN bootstrap for domain queries
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"https://rdap.org/domain/{quote(domain)}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        return self._parse_domain_response(data)
                    elif response.status == 404:
                        return {"domain": domain, "error": "Not found"}
                    else:
                        return {
                            "domain": domain,
                            "error": f"API error: {response.status}",
                        }
        except Exception as e:
            logger.error(f"Error querying RDAP for domain {domain}: {e}")
            return {"domain": domain, "error": str(e)}

    async def _query_rir(
        self, resource: str, rir: str, resource_type: str
    ) -> Dict[str, Any]:
        """
        Query specific RIR RDAP server.

        Args:
            resource: Resource to query (IP, ASN, etc.)
            rir: RIR name
            resource_type: Type of resource (ip, autnum, domain)

        Returns:
            Dictionary with registration data
        """
        try:
            base_url = self.BOOTSTRAP_SERVERS[rir]
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                url = f"{base_url}/{resource_type}/{quote(resource)}"
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        return self._parse_response(data, resource_type, rir)
                    elif response.status == 404:
                        return {"resource": resource, "rir": rir, "error": "Not found"}
                    else:
                        logger.warning(
                            f"RDAP query to {rir} returned {response.status}"
                        )
                        return {
                            "resource": resource,
                            "rir": rir,
                            "error": f"API error: {response.status}",
                        }
        except asyncio.TimeoutError:
            logger.error(f"Timeout querying RDAP for {resource} at {rir}")
            return {"resource": resource, "rir": rir, "error": "Query timeout"}
        except Exception as e:
            logger.error(
                f"Error querying RDAP for {resource} at {rir}: {e}", exc_info=True
            )
            return {"resource": resource, "rir": rir, "error": str(e)}

    def _parse_response(
        self, data: Dict[str, Any], resource_type: str, rir: str
    ) -> Dict[str, Any]:
        """Parse RDAP response based on resource type."""
        if resource_type == "ip":
            return self._parse_ip_response(data, rir)
        elif resource_type == "autnum":
            return self._parse_asn_response(data, rir)
        else:
            return data

    def _parse_ip_response(self, data: Dict[str, Any], rir: str) -> Dict[str, Any]:
        """Parse IP address RDAP response."""
        entities = []
        for entity in data.get("entities", []):
            entities.append(
                {
                    "handle": entity.get("handle"),
                    "name": self._get_entity_name(entity),
                    "roles": entity.get("roles", []),
                    "email": self._get_entity_email(entity),
                }
            )

        return {
            "start_address": data.get("startAddress"),
            "end_address": data.get("endAddress"),
            "ip_version": data.get("ipVersion"),
            "name": data.get("name"),
            "type": data.get("type"),
            "country": data.get("country"),
            "entities": entities,
            "status": data.get("status", []),
            "rir": rir,
            "handle": data.get("handle"),
            "registration_date": self._get_event_date(data, "registration"),
            "last_changed_date": self._get_event_date(data, "last changed"),
        }

    def _parse_asn_response(self, data: Dict[str, Any], rir: str) -> Dict[str, Any]:
        """Parse ASN RDAP response."""
        entities = []
        for entity in data.get("entities", []):
            entities.append(
                {
                    "handle": entity.get("handle"),
                    "name": self._get_entity_name(entity),
                    "roles": entity.get("roles", []),
                    "email": self._get_entity_email(entity),
                }
            )

        return {
            "asn": data.get("startAutnum"),
            "name": data.get("name"),
            "type": data.get("type"),
            "country": data.get("country"),
            "entities": entities,
            "status": data.get("status", []),
            "rir": rir,
            "handle": data.get("handle"),
            "registration_date": self._get_event_date(data, "registration"),
            "last_changed_date": self._get_event_date(data, "last changed"),
        }

    def _parse_domain_response(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Parse domain RDAP response."""
        nameservers = []
        for ns in data.get("nameservers", []):
            nameservers.append(ns.get("ldhName"))

        entities = []
        for entity in data.get("entities", []):
            entities.append(
                {
                    "handle": entity.get("handle"),
                    "name": self._get_entity_name(entity),
                    "roles": entity.get("roles", []),
                    "email": self._get_entity_email(entity),
                }
            )

        return {
            "domain": data.get("ldhName"),
            "unicode_name": data.get("unicodeName"),
            "status": data.get("status", []),
            "nameservers": nameservers,
            "entities": entities,
            "registration_date": self._get_event_date(data, "registration"),
            "expiration_date": self._get_event_date(data, "expiration"),
            "last_changed_date": self._get_event_date(data, "last changed"),
        }

    def _get_entity_name(self, entity: Dict[str, Any]) -> Optional[str]:
        """Extract entity name from vCard."""
        vcard_array = entity.get("vcardArray")
        if not vcard_array or len(vcard_array) < 2:
            return None

        for vcard_item in vcard_array[1]:
            if isinstance(vcard_item, list) and len(vcard_item) >= 4:
                if vcard_item[0] == "fn":
                    return vcard_item[3]
        return None

    def _get_entity_email(self, entity: Dict[str, Any]) -> Optional[str]:
        """Extract entity email from vCard."""
        vcard_array = entity.get("vcardArray")
        if not vcard_array or len(vcard_array) < 2:
            return None

        for vcard_item in vcard_array[1]:
            if isinstance(vcard_item, list) and len(vcard_item) >= 4:
                if vcard_item[0] == "email":
                    return vcard_item[3]
        return None

    def _get_event_date(self, data: Dict[str, Any], event_action: str) -> Optional[str]:
        """Extract event date from RDAP response."""
        events = data.get("events", [])
        for event in events:
            if event.get("eventAction") == event_action:
                return event.get("eventDate")
        return None


# Global client instance
_rdap_client: Optional[RDAPClient] = None


def get_rdap_client() -> RDAPClient:
    """Get or create the global RDAP client instance."""
    global _rdap_client
    if _rdap_client is None:
        _rdap_client = RDAPClient()
    return _rdap_client

File: go-backend/internal/datasources/rdap.go
package datasources

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type RDAPClient struct {
	httpClient *http.Client
}

var rdapBootstrapServers = map[string]string{
	"arin":    "https://rdap.arin.net/registry",
	"ripe":    "https://rdap.db.ripe.net",
	"apnic":   "https://rdap.apnic.net",
	"lacnic":  "https://rdap.lacnic.net/rdap",
	"afrinic": "https://rdap.afrinic.net/rdap",
}

func NewRDAPClient(timeout time.Duration) *RDAPClient {
	return &RDAPClient{
		httpClient: &http.Client{Timeout: timeout},
	}
}

func (c *RDAPClient) QueryIP(ctx context.Context, ipAddress string, rir string) map[string]any {
	if rir != "" {
		if base, ok := rdapBootstrapServers[rir]; ok {
			return c.queryRIR(ctx, ipAddress, rir, base, "ip")
		}
	}
	for rirName, base := range rdapBootstrapServers {
		result := c.queryRIR(ctx, ipAddress, rirName, base, "ip")
		if result["error"] == nil {
			return result
		}
	}
	return map[string]any{"ip": ipAddress, "error": "Not found in any RIR"}
}

func (c *RDAPClient) QueryASN(ctx context.Context, asn int64, rir string) map[string]any {
	resource := fmt.Sprintf("%d", asn)
	if rir != "" {
		if base, ok := rdapBootstrapServers[rir]; ok {
			return c.queryRIR(ctx, resource, rir, base, "autnum")
		}
	}
	for rirName, base := range rdapBootstrapServers {
		result := c.queryRIR(ctx, resource, rirName, base, "autnum")
		if result["error"] == nil {
			return result
		}
	}
	return map[string]any{"asn": asn, "error": "Not found in any RIR"}
}

func (c *RDAPClient) QueryDomain(ctx context.Context, domain string) map[string]any {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, "https://rdap.org/domain/"+url.PathEscape(domain), nil)
	if err != nil {
		return map[string]any{"domain": domain, "error": err.Error()}
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return map[string]any{"domain": domain, "error": err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return map[string]any{"domain": domain, "error": "Not found"}
	}
	if resp.StatusCode != http.StatusOK {
		return map[string]any{"domain": domain, "error": fmt.Sprintf("API error: %d", resp.StatusCode)}
	}

	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return map[string]any{"domain": domain, "error": err.Error()}
	}
	return parseDomainResponse(data)
}

func (c *RDAPClient) queryRIR(ctx context.Context, resource, rir, base, resourceType string) map[string]any {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, base+"/"+resourceType+"/"+url.PathEscape(resource), nil)
	if err != nil {
		return map[string]any{"resource": resource, "rir": rir, "error": err.Error()}
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return map[string]any{"resource": resource, "rir": rir, "error": err.Error()}
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return map[string]any{"resource": resource, "rir": rir, "error": "Not found"}
	}
	if resp.StatusCode != http.StatusOK {
		return map[string]any{"resource": resource, "rir": rir, "error": fmt.Sprintf("API error: %d", resp.StatusCode)}
	}

	var data map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return map[string]any{"resource": resource, "rir": rir, "error": err.Error()}
	}
	if resourceType == "ip" {
		return parseIPResponse(data, rir)
	}
	return parseASNResponse(data, rir)
}

func parseIPResponse(data map[string]any, rir string) map[string]any {
	return map[string]any{
		"start_address":     data["startAddress"],
		"end_address":       data["endAddress"],
		"ip_version":        data["ipVersion"],
		"name":              data["name"],
		"type":              data["type"],
		"country":           data["country"],
		"entities":          parseEntities(data),
		"status":            toAnySlice(data["status"]),
		"rir":               rir,
		"handle":            data["handle"],
		"registration_date": eventDate(data, "registration"),
		"last_changed_date": eventDate(data, "last changed"),
	}
}

func parseASNResponse(data map[string]any, rir string) map[string]any {
	return map[string]any{
		"asn":               data["startAutnum"],
		"name":              data["name"],
		"type":              data["type"],
		"country":           data["country"],
		"entities":          parseEntities(data),
		"status":            toAnySlice(data["status"]),
		"rir":               rir,
		"handle":            data["handle"],
		"registration_date": eventDate(data, "registration"),
		"last_changed_date": eventDate(data, "last changed"),
	}
}

func parseDomainResponse(data map[string]any) map[string]any {
	nameservers := make([]any, 0)
	for _, raw := range toAnySlice(data["nameservers"]) {
		if ns, ok := raw.(map[string]any); ok {
			nameservers = append(nameservers, ns["ldhName"])
		}
	}
	return map[string]any{
		"domain":            data["ldhName"],
		"unicode_name":      data["unicodeName"],
		"status":            toAnySlice(data["status"]),
		"nameservers":       nameservers,
		"entities":          parseEntities(data),
		"registration_date": eventDate(data, "registration"),
		"expiration_date":   eventDate(data, "expiration"),
		"last_changed_date": eventDate(data, "last changed"),
	}
}

func parseEntities(data map[string]any) []map[string]any {
	entities := make([]map[string]any, 0)
	for _, raw := range toAnySlice(data["entities"]) {
		entity, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		entities = append(entities, map[string]any{
			"handle": entity["handle"],
			"name":   vcardField(entity, "fn"),
			"roles":  toAnySlice(entity["roles"]),
			"email":  vcardField(entity, "email"),
		})
	}
	return entities
}

func vcardField(entity map[string]any, field string) any {
	vcardArray, ok := entity["vcardArray"].([]any)
	if !ok || len(vcardArray) < 2 {
		return nil
	}
	items, ok := vcardArray[1].([]any)
	if !ok {
		return nil
	}
	for _, raw := range items {
		entry, ok := raw.([]any)
		if !ok || len(entry) < 4 {
			continue
		}
		if key, ok := entry[0].(string); ok && key == field {
			return entry[3]
		}
	}
	return nil
}

func eventDate(data map[string]any, action string) any {
	for _, raw := range toAnySlice(data["events"]) {
		event, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		if event["eventAction"] == action {
			return event["eventDate"]
		}
	}
	return nil
}
